src/libcharon/plugins/kernel_super/kernel_super_ipsec.c

```diff
@@ -192,8 +192,133 @@ static status_t send_del_route_to_routes_conn(private_kernel_super_ipsec_t* this
+static host_t* _get_sp_lookup_addr(const kernel_ipsec_manage_policy_t* const data,
+								   const kernel_ipsec_policy_id_t* const id);
+
+static void write_tunnel_oper_items(private_kernel_super_ipsec_t* this,
+									kernel_ipsec_manage_policy_t* data,
+									kernel_ipsec_policy_id_t* id,
+									ike_super_peer_info_t* super_peer_info,
+									child_sa_t * child_sa,
+									bool write_uptime) {
+	if(IKE_SUPER_PEER_INFO_TYPE_TUNNEL == super_peer_info->get_type(super_peer_info)) {
+		return;
+	}
+	ike_sa_t* const ike_sa = child_sa->get_ike_sa(child_sa);
+	ike_sa_id_t* const ike_sa_id = ike_sa->get_id(ike_sa);
+	proposal_t* child_sa_proposal = child_sa->get_proposal(child_sa);
+	ike_cfg_t* curr_ike_cfg = ike_sa->get_ike_cfg(ike_sa);
+	struct terminator_tunnel_oper_data oper_data;
+
+	uint16_t enc_alg = ENCR_UNDEFINED, key_size, int_alg = AUTH_UNDEFINED;
+	uint16_t esn = NO_EXT_SEQ_NUMBERS;
+
+	if(! child_sa_proposal->get_algorithm(child_sa_proposal, ENCRYPTION_ALGORITHM, &enc_alg, &key_size)) {
+		LOG(DBG_KNL, SL_LOG_ERR, "encryption algorithm not found");
+		goto out;
+	}
+	if(! child_sa_proposal->get_algorithm(child_sa_proposal, INTEGRITY_ALGORITHM, &int_alg, NULL)) {
+		LOG(DBG_KNL, SL_LOG_ERR, "integrity algorithm not found");
+		goto out;
+	}
+	if(! child_sa_proposal->get_algorithm(child_sa_proposal, EXTENDED_SEQUENCE_NUMBERS, &esn, NULL)) {
+		LOG(DBG_KNL, SL_LOG_ERR, "extended sequence numbers not found");
+		goto out;
+	}
+
+	oper_data.encryption_algorithm = enc_alg;
+	oper_data.key_size = key_size;
+	oper_data.integrity_algorithm = int_alg;
+	oper_data.ext_seq_num = esn;
+	oper_data.dh_group = curr_ike_cfg->get_dh_group(curr_ike_cfg);
+	oper_data.anti_replay_window_size = child_sa->get_config(child_sa)->get_replay_window(child_sa->get_config(child_sa));
+	oper_data.version = ike_sa_id->get_ike_version(ike_sa_id);
+
+	oper_data.inbound_ipsec_spi = child_sa->get_spi(child_sa, TRUE);
+	oper_data.outbound_ipsec_spi = child_sa->get_spi(child_sa, FALSE);
+	oper_data.inbound_ike_spi = ike_sa_id->get_my_spi(ike_sa_id);
+	oper_data.outbound_ike_spi = ike_sa_id->get_other_spi(ike_sa_id);
+	oper_data.nat_identified = child_sa->has_encap(child_sa);
+	oper_data.status = ike_sa->get_state(ike_sa);
+	oper_data.activeness_status = ike_sa->get_is_active(ike_sa);
+
+	char device_public_addr[MAX_IP_STR_LEN];
+	char remote_public_addr[MAX_IP_STR_LEN];
+	char local_network[MAX_IP_PREFIX_STR_LEN];
+	if(! data->src->address_to_string(data->src, device_public_addr)) {
+		LOG(DBG_KNL, SL_LOG_ERR, "failed to get device public address");
+		goto out;
+	}
+	if(! data->src->address_to_string(data->dst, remote_public_addr)) {
+		LOG(DBG_KNL, SL_LOG_ERR, "failed to get remote public address");
+		goto out;
+	}
+
+	struct pm_proto_prefix prefix;
+	if(! ts_to_picomsg_prefix(id->dst_ts, &prefix)) {
+		LOG(DBG_KNL, SL_LOG_ERR, "failed to get local network");
+		goto out;
+	}
+	if(! pm_proto_addr_to_string(&prefix.prefix, local_network)) {
```

Comment:
I suggest you add a new `ts_to_prefix_string()` function which will call these 2 functions (similar to `ts_to_prefix_string_for_logs()` but returns a bool).
This logic seems useful outside the context of this function.